### PR TITLE
fix Issue 15089 - Marks wrong line as where error occurs.

### PR DESCRIPTION
**Pull request by WalterBright**
https://issues.dlang.org/show_bug.cgi?id=15089

This just adds a `loc` parameter to get the right location to a bunch of constant folding routines.

### CODE REVIEW

**Comment by andralex** [comment]
noice

**Comment by yebblies** [comment]
IMO it looks like the conversion checking code was just using the loc of the wrong expression, but this is good to have anyway.

File: test/fail_compilation/fail15089.d [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/fail15089.d(10): Error: cannot implicitly convert expression (130) of type int to byte
```
Comment (jacob-carlborg) [inline test/fail_compilation/fail15089.d]: I see that the error message now points to the correct line. But unfortunately it still prints an expression that's not in the source code, `130`. What happens if I do something stupid like putting two expressions on the same line that result in the same value:

```d
byte piece = Pieces.Rook ^ color; byte piece2 = 129 + 1;
```

Ideally it should print both the original expression from the source code and that value it was evaluated to.

Comment (WalterBright) [inline test/fail_compilation/fail15089.d]: If you put two expressions on the same line, and are confused about which one is the error, I'd edit the file and put them on separate lines.

After all the AST rewrites the compiler goes through, getting the original back can be problematic beyond just pointing to the source line.

Comment (jacob-carlborg) [inline test/fail_compilation/fail15089.d]: If the start and end of the expression is stored the compiler can just read the original buffer.

Comment (yebblies) [inline test/fail_compilation/fail15089.d]: Quite often printing the original expression is less helpful.

**Pull request merged**